src/logic/gcs.js

```diff
@@ -0,0 +1,87 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+// @flow
+
+import { Storage } from '@google-cloud/storage';
+import { Writable } from 'stream';
+
+import { getLogger } from '../log';
+
+interface GcsStorage {
+  ping(): Promise<void>;
+  getWriteStreamForFile(filePath: string): Writable;
+}
+
+class RealGcsStorage implements GcsStorage {
+  bucket: any;
+
+  constructor(bucket: any) {
+    this.bucket = bucket;
+  }
+
+  async ping(): Promise<void> {
+    const [result] = await this.bucket.exists();
+    if (!result) {
+      throw new Error(`The bucket ${this.bucket.name} doesn't exist.`);
+    }
+  }
+
+  getWriteStreamForFile(filePath: string): Writable {
+    const file = this.bucket.file(filePath);
+    const googleStorageStream = file.createWriteStream({
+      public: true, // Our content is already gzipped
+      resumable: false,
+      gzip: false,
+      metadata: {
+        contentType: 'text/plain',
+        cacheControl: 'max-age: 365000000, immutable',
+        contentEncoding: 'gzip',
+      },
+    });
+    return googleStorageStream;
+  }
+}
+
+class MockGcsStorage implements GcsStorage {
+  ping() {
+    return Promise.resolve();
+  }
+
+  getWriteStreamForFile() {
+    // We create a Writable stream that just consumes everything.
+    const sinkWriteStream = new Writable({
+      write(chunk, encoding, callback) {
+        callback();
+      },
+    });
+    return sinkWriteStream;
+  }
+}
+
+type GcsConfig = {
+  +googleAuthenticationFilePath: string,
+  +gcsBucket: string,
+};
+
+export function create(config: GcsConfig): GcsStorage {
+  const log = getLogger('gcs.create');
+  const googleAuthenticationFilePath = config.googleAuthenticationFilePath;
+  if (googleAuthenticationFilePath === 'MOCKED') {
+    log.debug('gcs_mocked', 'Returning the mocked Storage backend.');
+    return new MockGcsStorage();
+  }
+
+  const gcsBucket = config.gcsBucket;
+  if (!gcsBucket) {
+    throw new Error('The bucket name should not be empty');
+  }
+
+  log.debug('gcs_real', 'Returning the real Storage backend.');
+  const storage = new Storage({
+    keyFilename: googleAuthenticationFilePath,
+  });
```

Comment:
This is what I found to be the easiest to locally configure Google Storage.
If we don't specify anything here, the library tries to find the information in various places:
* an environment variable `GOOGLE_APPLICATION_CREDENTIALS` can also hold this file path
* the library also checks a so called "well known path", in `$HOME/.config/gcloud/application_default_credentials.json` (replace `$HOME/.config` by `APPDATA` in windows).
* it also checks if it's in GCE (google computing engine) and can get its credentials from there too.

I like none of these possibilities for the dev environment, that's why instead I wrote the code to explicitely specify it. A previous version of the code was setting the environment variable `GOOGLE_APPLICATION_CREDENTIALS`, but I thought it was better like this.
For the production environment, it's still not clear to me. I've been told that it "just works" with environment variables, so my guess is that if `googleAuthenticationFilePath` stays empty (its default value) this will work like that. There's also a chance that I'll have to come back and change this later.

Reply:
I'm confused by the local config implementation. In other projects I've used a `.ENV` file that I can locally set environment variables to. The `local-config.json` feels odd to me in this case. Can you remind me what the decision and thinking was for this design?

Reply:
Environment variables are how this should be configured in the production environment.
I decided to use a `local-config.json` file for the following reasons:
* this is close to the "local configuration" mechanism from the profiler's server.js (file `webpack.local-config.js`)
* the configuration library `convict` makes it easy.

Though I also like your idea of using a `.env` file. How would you implement it? I see there's https://github.com/toddbluhm/env-cmd and I could add it to the `start` script command, what do you think?

Reply:
I've used [dotenv](https://www.npmjs.com/package/dotenv) in the past. I think there is a benefit to following having a `.ENV` file, as it's a fairly common practice. However, I'm happy to go with what you think is best, especially if the configuration library provides some additional safety checks.

Reply:
Thanks, I used dotenv to implement this.